detox/src/configuration/composeRunnerConfig.js

```diff
@@ -1,19 +1,131 @@
-// @ts-nocheck
+const os = require('os');
+
+const _ = require('lodash');
+
+const log = require('../utils/logger');
+
+/**
+ * @param {object} opts
+ * @param {Detox.DetoxConfig} opts.globalConfig
+ * @param {Detox.DetoxConfiguration} opts.localConfig
+ * @param {DetoxInternals.CLIConfig} opts.cliConfig
+ * @param {Record<string, any>} opts.testRunnerArgv
+ * @param {import('../errors/DetoxConfigErrorComposer')} opts.errorComposer
+ * @returns {Detox.DetoxTestRunnerConfig} opts.testRunnerArgv
+ */
+function composeRunnerConfig(opts) {
+  const globalConfig = adaptLegacyRunnerConfig(opts.globalConfig);
+  if (globalConfig != null && typeof globalConfig !== 'object') {
+    throw opts.errorComposer.invalidTestRunnerProperty(true);
+  }
+
+  const localConfig = opts.localConfig.testRunner;
+  if (localConfig != null && typeof localConfig !== 'object') {
+    throw opts.errorComposer.invalidTestRunnerProperty(false);
+  }
+
+  const cliConfig = opts.cliConfig;
+
+  /** @type {Detox.DetoxTestRunnerConfig} */
+  const merged = _.merge(
+    {
+      retries: 0,
+      inspectBrk: inspectBrkHookDefault,
+      forwardEnv: false,
+      bail: false,
+      jest: {
+        setupTimeout: 300000,
+        teardownTimeout: 30000,
+        retryAfterCircusRetries: false,
+        reportSpecs: undefined,
+        reportWorkerAssign: true,
+      },
+      args: {
+        $0: 'jest',
+        _: [],
+      },
+    },
+    globalConfig,
+    localConfig,
+    cliConfig.retries != null ? { retries: cliConfig.retries } : null,
+    cliConfig.jestReportSpecs != null ? { jest: { reportSpecs: cliConfig.jestReportSpecs } } : null,
+    {
+      args: _.omitBy(opts.testRunnerArgv, hasEmptyPositionalArgs)
+    }
+  );
+
+  if (typeof merged.inspectBrk === 'function') {
+    if (cliConfig.inspectBrk) {
+      merged.retries = 0;
+      merged.forwardEnv = true;
+      merged.inspectBrk(merged);
+    }
+
+    delete merged.inspectBrk;
+  }
+
+  return merged;
+}
+
+function adaptLegacyRunnerConfig(globalConfig) {
+  let isLegacy = false;
+
+  const runnerConfigKey = 'runnerConfig' in globalConfig ? 'runnerConfig' : 'runner-config';
+  if (_.isString(globalConfig[runnerConfigKey])) {
+    isLegacy = true;
+    log.warn(`Detected a deprecated "${runnerConfigKey}" property (string).`);
+  }
+
+  const testRunnerKey = 'testRunner' in globalConfig ? 'testRunner' : 'test-runner';
+  if (_.isString(globalConfig[testRunnerKey])) {
+    isLegacy = true;
+    log.warn(`Detected a deprecated "${testRunnerKey}" property (string).`);
+  }
+
+  if (globalConfig.specs != null) {
+    isLegacy = true;
+    log.warn(`Detected a deprecated "specs" property.`);
+  }
+
+  if (!isLegacy) {
+    return globalConfig.testRunner;
+  }
+
+  log.warn(`Please migrate your Detox config according to the guide:\nhttps://wix.github.io/Detox/docs/next/guide/migration\n`);
```

Comment:
after the release it won't be `wix.github.io/Detox/docs/next/guide/migration` but `wix.github.io/Detox/docs/guide/migration`

Reply:
Of course! This was intended!
Let's just add a TODO item to `Replace All` before merging.

Reply:
Added a TODO.